### PR TITLE
Better creative paging buttons

### PR DESCRIPTION
**Pull request by ExtraCrafTX**
new-buttons

### CODE REVIEW

**Comment by modmuss50** [comment]
How does it look when the top row of tabs are full? I assume you left that gap there?

**Comment by Prospector** [comment]
That gap's already always there

**Comment by shedaniel** [comment]
iirc the top row can never be full. Correct me if I am wrong

**Comment by ExtraCrafTX** [comment]
It cannot be full, the gap is always there because it's too small to properly fit another tab.

**Comment by RedstoneParadox** [comment]
They still seem too small to me. Also, you have to hover over them to see the current page.

**Comment by i509VCB** [comment]
Well I think the buttons are a bit small but have we considered possibly allowing for shrinking the toolbars button to expand the page buttons to be larger.

**Comment by shedaniel** [comment]
I don’t like the idea of editing existing vanilla stuff just to add a button

**Comment by LemmaEOF** [comment]
Yeah, editing existing vanilla stuff is out of scope for API.  Maybe these buttons could be a bit taller? It'd probably look really weird to have them as tall as a regular tab, but maybe half-tab?

**Comment by MarcusElg** [comment]
Have you tried with taller buttons?

**Comment by ghost** [comment]
I like the button size, but liked them when they were detached like you showed in this image:  
But honestly it would be really great if it was configurable in terms of position. Then people would be able to move them down to have them attached to the creative menu or move them over to the side like old buttons where.

**Comment by modmuss50** [comment]
Can you just bump the version number of the creative tabs module?

**Comment by oliviathevampire** [comment]
What if we can add the different variants from #326 as different config options?

**Comment by modmuss50** [comment]
No, I dont think there is a need for 2 button styles in fabric. Someone is free to make a mod that adds more styles, or even to redo the whole gui.

**Comment by ghost** [comment]
Then why even allow this pr to go ahead when it can be a mod?

**Comment by modmuss50** [comment]
This PR is a simple improvment to the existing buttons, its not adding any new features or larger scale changes.

**Comment by kgscialdone** [comment]
Creative tabs fall within the 90% rule. There needs to be some way to extend the available space, thus the pages. Editing the page switcher buttons or having multiple types that you can choose between is probably outside the 90% rule, though.

**Comment by ghost** [comment]
Changing its position though can be a simple properties files which fabric api already contains one of...

**Comment by modmuss50** [comment]
I dont see a big reason why a user may want to move the buttons?
A user is more likely going to install a mod that overhalls the entire gui than tweak a config to move the buttons a few pixels.

**Comment by ghost** [comment]
Well lets say not all users aren't going to like the creative buttons moving to up between the tabs.

**Comment by modmuss50** [comment]
Then tough IMO, Its a very minor thing that is only very slightly different from what is already have currently.

I see no reason why the user would want to move the buttons to their old place because they dont like this new place.

**Comment by ghost** [comment]
> because they dont like this new place.

**Comment by modmuss50** [comment]
I give up with this argument, the new place imo is no better or worse than the old place. A lot of people in this thread seem to prefer the new buttons over the old buttons, if more people dont like the new place then possibly go back to #326 ? or leave it be. I dont see that happening tho.

Sorry for spamming people with emails if they have them enabled.

**Comment by ExtraCrafTX** [comment]
Ok, in terms of making them taller as tabs, I think they already look kinda weird even as half-tabs:
as-tab-longer

In terms of making them detached, I do like them better, although I'm not sure what most people think of those. I kinda just went with tabs because they had the most positive feedback in #326

In terms of bumping the version, do I just change everywhere it says v0 to v1? And do I bump the patch number here: `version = getSubprojectVersion(project, "0.1.0")`?

**Comment by modmuss50** [comment]
No, leave it as v0. Just bump the patch number in the build.gradle.

Personally I like them attached.

**Comment by MarcusElg** [comment]
Defently attached! Do prefer the taller ones tho.

**Comment by Prospector** [comment]
@sfPlayer1 where would they go to be more intuitive while also not being island buttons outside of the interface?

**Comment by Player3324** [comment]
I think they have to be outside to look better, like a vertical up/down on the left side to cycle between content tab sets. The "island" idea doesn't apply, there's no natural flow to start with.

**Comment by falkreon** [comment]
I don't see the old location or the new location as intuitive locations at all. like it or not, the forge locations felt intuitive to me. if I had to pick between just the two, I'd reject this PR and use the current locations.

**Comment by liach** [comment]
The buttons are just too small to me. Will have a hard time clicking them.

**Comment by shedaniel** [comment]
image
This but bigger buttons? It looks weird imo

**Comment by Sollace** [comment]
@shedanial Wings! But, uh, no. I wouldn't attach them to the sides of the tabs like that.

**Comment by Sollace** [comment]
What if you moved them down (closer to the vertical center) and made them taller? A bit like the buttons we used to have on XBox edition.

64491065-64d32f80-d296-11e9-93f5-8a20cbdcf6cf

(forgive the horrible font)

Edit: Not XBox edition, Win10 edition. It's the long tab that flies out the craft book, as I remember.

**Comment by Player3324** [comment]
I suggested vertical on the left since it makes more sense with the 3 sticky tabs on the right (only the left part of the screen scrolls through, which physically implies a rotation around a horizontal axis with vertical travel).

Sollace's left+right buttons may be a bit annoying from the long mouse travel distance when switching directions when correcting over-scroll, but they certainly looks nice.

**Comment by Sollace** [comment]
@sfPlayer1 Note that we could always include a keyboard shortcut for "Power Players". I will admit the tracking time may be troublesome when switching rapidly between them, though likely not more than switching between the top and bottom tabs.

**Comment by shedaniel** [comment]
I guess this should be rebased on 1.15 branch.

**Comment by falkreon** [comment]
I still hate the snapped-together buttons

**They do a poor job of communicating intent**. What do they scroll left and right through? Normally there's an affordance that scroll buttons are around either the Thing that they scroll, or a model of the Thing (a scrollbar being one such model). Similarly, I'll go looking for buttons in places that communicate some kind of intent, later finding these and going "maybe this'll do it". That's not good.

**They don't really help the mouse travel "problem"**. No matter where you put them, you're going to need to move across half the UI on average to pick up a block. I also find overscroll to be rare, and worth the extra travel for an exceptional circumstance. The deeper problem here is having too many tabs, and no good way for mods to band together to push categories like vanilla. You could be making one or two clicks to go to Generators, then scrolling down to find the right one. But that's not really for us to solve here, is it?

**They cause extra misclick problems**, at least for me. Having them smashed together means I'm likely to start going the wrong direction through pages, requiring the very same corrective clicks they're meant to streamline. I imagine it's even harder on a tablet. Please, please, make big buttons for my fat mouse-fingers. Those tall ones look excellent!

**There's a better way to solve the problem already mentioned in-thread**, by enabling arrow-key scrolling.

**Comment by ExtraCrafTX** [comment]
> **They do a poor job of communicating intent**

I don't think there is a particularly good way to communicate intent in this case since the tabs weren't really meant to be scrolled in the first place.

> **They don't really help the mouse travel "problem"**

I somewhat agree with this. They do solve the problem in the sense that you don't need to travel across the screen to press the opposite button, but yes, it doesn't change the fact that mouse travel is still required for items.  
I don't think the point was that the current buttons solve the travel problem, more that the separated design would introduce more travel.

> **They cause extra misclick problems**

Fair enough, I agree that the tall ones that Sollace sent are much better for this. I feel like the main problem with those is that they're thin and separated quite a lot. That does make the travel problem a lot worse.  
I also find that position to not be as clear as the position in this PR, but that's likely just me being biased.

I definitely agree that it would be better if the buttons could be made bigger though.

> **There's a better way to solve the problem already mentioned in-thread**

I don't know about that. I mean yes, adding keybinds for moving between tab pages kind of solves the problem. But really, the problem has just moved if we're doing arrow keys, because now I need to move my hands off the mouse to the arrow keys, and then back. That to me is just as bad as the travel problem.

I don't know that its helping that the request is repeatedly getting delayed for these discussions. I mean obviously these buttons are not the end goal, the point was just that its a tiny change to look and feel slightly better than the current buttons which, I think we can all agree, are really not good. Once we have somewhat of a better idea as to how to do the buttons, that can be a separate change, no?

**Comment by falkreon** [comment]
But is this an improvement? Moving the arrows inside the tabs makes the potential for misclicks worse, and now there'll be no room to make the buttons larger, which is the real first problem here.

My point with mouse travel - you've got a good point that hand travel needs to happen for the keyboard unless tab scrolling is on A and D - my point is that once you're scrolling, if you overscroll on arrow keys, you've got virtually no travel to correct that overscroll.

**Comment by ExtraCrafTX** [comment]
> there'll be no room to make the buttons larger

There already isn't any room for that. These buttons may not be a usability improvement, but the current buttons just look awful to me and they don't have any hover effect, for example.

> you've got virtually no travel to correct that overscroll

Yeah, that is true. Although again, just having the buttons together already solves that, and removes hand travel, but does have more of a misclick chance than separated buttons as you mentioned.  
Either way, for this the buttons in this PR are no worse than the current ones in my opinion.

And as I mentioned, this PR was just meant to be an aesthetic change, and people seemed mostly positive about the new position, so I settled on the new one. If merging this requires going back to the old position as shown in the first image [here](https://github.com/FabricMC/fabric/pull/326#issuecomment-518297495) in my message in #326, I am more than willing to do that.

**Comment by shedaniel** [comment]
is it possible to have this updated to 1.15 button highlight

**Comment by Prospector** [comment]
This style of button still has the blue highlight in vanilla (see recipe book icon), though the one currently in Fabric API (which I believe has no highlight currently) should use a white one

**Comment by shedaniel** [comment]
we should switch to white outline. since the recipe book button is an issue: https://bugs.mojang.com/browse/MC-165979

**Comment by Prospector** [comment]
What would it even look like since the button already has a partially white outline to begin with? I'd say wait for some vanilla precedent and until then, match vanilla.

**Comment by shedaniel** [comment]
Why do we need to wait for vanilla? The recipe book issue is not related to our tab buttons, it is not like our tab buttons coexist with the vanilla recipe book buttons anyways, creative inventory doesn't show the recipe book button, also the "partially white outline" is in button as well, so that is basically invalid.

**Comment by Prospector** [comment]
Not sure what you mean, normal buttons are grey with a black outline that turns to white when hovering. Buttons in this style like the recipe book button are themed as in-game screen backgrounds and have white highlights, which would mix with the white border (would probably look very strange if the black border was simply turned white). Every other type of button in this style has a blue highlight so that's what fabric api should use imo

**Comment by shedaniel** [comment]
Ah right you are correct.

File: fabric-item-groups-v0/src/main/java/net/fabricmc/fabric/mixin/item/group/client/MixinCreativePlayerInventoryGui.java [hunk 1]
```diff
@@ -120,9 +120,9 @@ private void init(CallbackInfo info) {
 		fabric_updateSelection();
 
 		int xpos = x + 170;
```
Comment (shedaniel) [inline fabric-item-groups-v0/src/main/java/net/fabricmc/fabric/mixin/item/group/client/MixinCreativePlayerInventoryGui.java:122]: I see no code here to change the button positions, I forked this and what I got is this:
image

Comment (ExtraCrafTX) [reply fabric-item-groups-v0/src/main/java/net/fabricmc/fabric/mixin/item/group/client/MixinCreativePlayerInventoryGui.java:122]: Huh. Probably lost that change during the rebase. I'll fix it and push soon.

Comment (shedaniel) [reply fabric-item-groups-v0/src/main/java/net/fabricmc/fabric/mixin/item/group/client/MixinCreativePlayerInventoryGui.java:122]: Please do.

**Comment by i509VCB** [comment]
Last comment about losing button positions during rebase needs to be resolved.

**Pull request closed without merging**